deepinv/distrib/distrib_framework.py

```diff
@@ -0,0 +1,1431 @@
+                raise RuntimeError(
+                    "GPU mode requested but CUDA not available or no visible GPUs"
+                )
+            if visible_gpus == 1:
+                # GPU isolation is handled externally - use the only visible GPU
+                dev_index = 0
+            else:
+                # Multiple GPUs visible - map local_rank to device index
+                dev_index = self.local_rank % visible_gpus
+            self.device = torch.device(f"cuda:{dev_index}")
+            torch.cuda.set_device(self.device)
+        else:
+            # Auto mode
+            if torch.cuda.is_available() and visible_gpus > 0:
+                # When CUDA_VISIBLE_DEVICES is set externally (e.g., by SLURM/submitit),
+                # each process sees only its assigned GPU(s). In this case, if there's
+                # only 1 visible GPU per process, just use cuda:0 for all processes.
+                # Otherwise, map local_rank onto visible devices.
+                if visible_gpus == 1:
+                    # GPU isolation is handled externally - use the only visible GPU
+                    dev_index = 0
+                else:
+                    # Multiple GPUs visible - map local_rank to device index
+                    dev_index = self.local_rank % visible_gpus
+
+                self.device = torch.device(f"cuda:{dev_index}")
+                torch.cuda.set_device(self.device)
+            else:
+                self.device = torch.device("cpu")
+
+        self._post_init_setup()
+        return self
+
+    def __exit__(self, exc_type, exc, tb):
+        # Only destroy process group if:
+        # 1. cleanup=True (caller wants cleanup)
+        # 2. We initialized it (initialized_here=True)
+        # 3. It's still initialized
+        if self.cleanup and self.initialized_here and dist.is_initialized():
+            try:
+                dist.barrier()
+            except Exception:
+                pass
+            dist.destroy_process_group()
+
+    # ----------------------
+    # Post-init knobs
+    # ----------------------
+    def _post_init_setup(self):
+        # Seeding (rank offset for reproducible-but-unique RNG per process)
+        if self.seed is not None:
+            s = self.seed + (self.rank if self.is_dist else 0)
+            torch.manual_seed(s)
+            if torch.cuda.is_available():
+                torch.cuda.manual_seed_all(s)
+
+        # Deterministic cuDNN
+        if self.deterministic:
+            torch.backends.cudnn.benchmark = False
+            torch.backends.cudnn.deterministic = True
+
+    # ----------------------
+    # Sharding
+    # ----------------------
+    def local_indices(self, num_items: int) -> list[int]:
+        r"""
+        Get local indices for this rank based on round robin sharding.
+
+        :param int num_items: total number of items to shard.
+        :return: (list) list of indices assigned to this rank.
```

Comment:
sphinx reads the return type from the type hinting (`-> list[int]`)
```suggestion
        :return: list of indices assigned to this rank.
```